### PR TITLE
[batch] Python jobs

### PR DESCRIPTION
**Pull request by jigold**

### CODE REVIEW

**Review by danking — CHANGES_REQUESTED** [review]

File: hail/python/hailtop/batch/docs/cookbook/files/run_rf_checkpoint.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+    # store obs and pred values for this window
+    obs = y_test["oe"].to_list()[0]
+    pred = y_pred[0]
+
+    return (window_name, obs, pred)
+
+
+def as_tsv(input: Tuple[str, float, float]) -> str:
+    return '\t'.join([str(i) for i in input])
```
Comment (daniel-goldstein) [inline hail/python/hailtop/batch/docs/cookbook/files/run_rf_checkpoint.py:41]: nit: don't need the `[` `]`

File: hail/python/hailtop/batch/job.py [hunk 2]
```diff
@@ -583,3 +374,542 @@ def _pretty(self):
...
+
+    def _get_resource(self, item: str) -> '_resource.Resource':
+        if item not in self._resources:
+            r = self._batch._new_job_resource_file(self, value=item)
+            self._resources[item] = r
+            self._resources_inverse[r] = item
+
+        return self._resources[item]
+
+    def declare_resource_group(self, **mappings: Dict[str, Any]) -> 'Job':
```
Comment (daniel-goldstein) [inline hail/python/hailtop/batch/job.py:417]: Should this return `BashJob`?

File: hail/python/hailtop/batch/job.py [hunk 3]
```diff
@@ -583,3 +374,542 @@ def _pretty(self):
...
+        for name, d in mappings.items():
+            assert name not in self._resources
+            if not isinstance(d, dict):
+                raise BatchException(f"value for name '{name}' is not a dict. Found '{type(d)}' instead.")
+            rg = self._batch._new_resource_group(self, d, root=name)
+            self._resources[name] = rg
+            _add_resource_to_set(self._valid, rg)
+        return self
+
+    def image(self, image: str) -> 'Job':
```
Comment (daniel-goldstein) [inline hail/python/hailtop/batch/job.py:465]: same `BashJob`

File: hail/python/hailtop/batch/job.py [hunk 4]
```diff
@@ -583,3 +374,542 @@ def _pretty(self):
 
     def __str__(self):
         return self._uid
+
+
+class BashJob(Job):
+    """
+    Object representing a single job to execute.
```
Comment (daniel-goldstein) [inline hail/python/hailtop/batch/job.py:381]: Should be specific to `BashJob`.

File: hail/python/hailtop/batch/job.py [hunk 5]
```diff
@@ -583,3 +374,542 @@ def _pretty(self):
...
+        subst_command = re.sub('(' + ')|('.join(regexes) + ')',
+                               handler,
+                               command)
+        self._command.append(subst_command)
+        return self
+
+
+class PythonJob(Job):
+    """
+    Object representing a single job to execute.
```
Comment (daniel-goldstein) [inline hail/python/hailtop/batch/job.py:612]: Should be specific to `PythonJob`.

File: hail/python/hailtop/batch/job.py [hunk 6]
```diff
@@ -583,3 +374,542 @@ def _pretty(self):
...
+        b = Batch(default_python_image='gcr.io/hail-vdc/python-dill:3.7-slim')
+
+        def multiply(x, y):
+            return x * y
+
+        def add(x, y):
+            return x + y
+
+        j = b.new_python_job()
+        j.call(multiply, 2, 3)
```
Comment (daniel-goldstein) [inline hail/python/hailtop/batch/job.py:632]: Should this be

```
result = j.call(multiply, 2, 3)
result = j.call(add, result, 3)
```

File: hail/python/hailtop/batch/job.py [hunk 7]
```diff
@@ -583,3 +374,542 @@ def _pretty(self):
...
+
+    The image to use for Python jobs can be specified by `default_python_image`
+    when constructing a :class:`.Batch`. The image specified must have the `dill`
+    package installed. If ``default_python_image`` is not specified, then a Docker
+    image will automatically be created for you with the base image
+    `hailgenetics/python-dill:[major_version].[minor_version]-slim` and the Python
+    packages specified by ``python_requirements`` will be installed. The default name
+    of the image is `batch-python` with a random string for the tag unless ``python_build_image_name``
+    is specified. If the :class:`.ServiceBackend` is the backend, the locally built
+    image will be pushed to the repository specified by ``image_repository``.
```
Comment (daniel-goldstein) [inline hail/python/hailtop/batch/job.py:654]: These aren't fields in `PythonJob`, so this explanation might be better placed on `Batch` or somewhere adjacent.

**Review by daniel-goldstein — CHANGES_REQUESTED** [review]
Very cool! Mostly my comments/concerns are around the docker image interface/steps. If there are areas of the python compilation that you want me to take a closer look at I can but it looks pretty good to me on this first pass.

File: hail/python/hailtop/batch/job.py (hunk 7 above)
Comment (jigold) [reply hail/python/hailtop/batch/job.py:654]: Moved to `Batch.new_python_job`

**Comment by danking** [comment]
@daniel-goldstein I'll let you complete the PR review on this, it looks fine to me

File: hail/python/hailtop/batch/docker.py [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+
+            sync_check_shell_output(f'docker build -t {fullname} {docker_path}')
+            print(f'finished building image {fullname}')
+        else:
+            sync_check_shell_output(f'docker pull {base_image}')
+            sync_check_shell_output(f'docker tag {base_image} {fullname}')
+            print(f'finished pulling image {fullname}')
+
+        image_split = fullname.rsplit('/', 1)
+        if len(image_split) == 2:
```
Comment (daniel-goldstein) [inline hail/python/hailtop/batch/docker.py:89]: I think `if '/' in fullname` is equivalent

**Review by daniel-goldstein — DISMISSED** [review]
I like this a lot better. And I see a near future where `build_python_image` can become a job that creates an image dependency. A much simpler interface indeed.

**Review by daniel-goldstein — APPROVED** [review]

**Pull request merged**